sky/serve/controller.py

```diff
@@ -0,0 +1,159 @@
+
+        @self.app.get('/controller/get_latest_info')
+        def get_latest_info():
+            latest_info = {
+                'replica_info':
+                    self.infra_provider.get_replica_info(verbose=True),
+                'uptime': self.infra_provider.get_uptime(),
+            }
+            latest_info = {
+                k: base64.b64encode(pickle.dumps(v)).decode('utf-8')
+                for k, v in latest_info.items()
+            }
+            return latest_info
+
+        @self.app.post('/controller/terminate')
+        def terminate(request: fastapi.Request):
+            del request
+            logger.info('Terminating service...')
+            if self.autoscaler is not None:
+                logger.info('Terminate autoscaler...')
+                self.autoscaler.terminate()
+            msg = self.infra_provider.terminate()
+            return {'message': msg}
+
+        # Run replica_prober and autoscaler (if autoscaler is defined)
+        # in separate threads in the background.
+        # This should not block the main thread.
+        self.infra_provider.start_replica_prober()
+        if self.autoscaler is not None:
+            self.autoscaler.start()
+
+        # Disable all GET logs if SKYPILOT_DEBUG is not set to avoid overflowing
+        # the controller logs.
+        if not env_options.Options.SHOW_DEBUG_INFO.get():
+            logging.getLogger('uvicorn.access').addFilter(
+                SuppressSuccessGetAccessLogsFilter())
+
+        logger.info(
+            f'SkyServe Controller started on http://localhost:{self.port}')
+        uvicorn.run(self.app, host='localhost', port=self.port)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='SkyServe Controller')
+    parser.add_argument('--service-name',
+                        type=str,
+                        help='Name of the service',
+                        required=True)
+    parser.add_argument('--task-yaml',
+                        type=str,
+                        help='Task YAML file',
+                        required=True)
+    parser.add_argument('--port',
+                        '-p',
+                        type=int,
+                        help='Port to run the controller',
+                        required=True)
+    args = parser.parse_args()
+
+    # Generate ssh key pair to avoid race condition when multiple sky.launch
+    # are executed at the same time.
+    authentication.get_or_generate_keys()
+
+    # ======= Infra Provider =========
+    service_spec = serve.SkyServiceSpec.from_yaml(args.task_yaml)
+    _infra_provider = infra_providers.SkyPilotInfraProvider(
+        args.task_yaml,
+        args.service_name,
+        readiness_suffix=service_spec.readiness_suffix,
+        initial_delay_seconds=service_spec.initial_delay_seconds,
+        post_data=service_spec.post_data)
+
+    # ======= Autoscaler =========
+    _autoscaler = autoscalers.RequestRateAutoscaler(
```

Comment:
To discuss: Is this tested in v0? If not, shall we consider leaving the entire autoscalers.py out?